### PR TITLE
kubelet: storage: teardown terminated pod volumes

### LINKED ISSUE #35406: EmptyDir can't be removed when pod become evicted
**Description of problem**:
Mount a emptyDir in pod `/tmp` then create large file in emptyDir to trigger nodefs.available eviction, when pod become Evicted, the large file can't be removed, node always keep in DiskPressure 

**Kubernetes version** (use `kubectl version`):
v1.5.0-alpha.1.804+a7db9bccb56abf-dirty

**What happened**:
File in emptyDir can't be removed after pod become Evicted, and node always keep in DiskPressure=True

**What you expected to happen**:
File in emptyDir should be removed after pod become Evicted, and node become DiskPressure=false after a period.

**How to reproduce it** (as minimally and precisely as possible):

```
1. configure kubelet with eviction-soft
--eviction-soft="nodefs.available<3Gi"
--eviction-soft-grace-period="nodefs.available=20s"
--eviction-pressure-transition-period=30s

2. Create a pod with emptyDir
$ kubectl create -f https://raw.githubusercontent.com/mdshuai/testfile-openshift/master/k8s/hello-pod.yaml

3. Create a large file in pod /tmp dir
$ kubectl exec hello-pod -- dd if=/dev/zero of=/tmp/test1 bs=10M count=1024

4. Wait pod become Evicted then check node status and file in emptyDir
[root@ip-172-18-6-200 kubernetes]# kubectl get pod
NAME        READY     STATUS    RESTARTS   AGE
hello-pod   0/1       Evicted   0          23m
[root@ip-172-18-6-200 kubernetes]# kubectl describe node 127.0.0.1|grep DiskPressure
  DiskPressure      True    Mon, 24 Oct 2016 02:56:22 -0400     Mon, 24 Oct 2016 02:33:01 -0400     KubeletHasDiskPressure      kubelet has disk pressure
  29m       29m     2   {kubelet 127.0.0.1}         Normal      NodeHasNoDiskPressure   Node 127.0.0.1 status is now: NodeHasNoDiskPressure
  23m       23m     1   {kubelet 127.0.0.1}         Normal      NodeHasDiskPressure Node 127.0.0.1 status is now: NodeHasDiskPressure
[root@ip-172-18-6-200 kubernetes]# ls -lh /var/lib/kubelet/pods/8095536f-99b3-11e6-9bc0-0ebd788c9e1f/volumes/kubernetes.io~empty-dir/tmp/test1
-rw-r--r--. 1 root root 5.3G Oct 24 02:33 /var/lib/kubelet/pods/8095536f-99b3-11e6-9bc0-0ebd788c9e1f/volumes/kubernetes.io~empty-dir/tmp/test1

```

**Anything else do we need to know**:
When delete the pod the large file in emptyDir then removed.

### PR DESCRIPTION
**Pull request by sjenning**
This is a continuation of the work done in https://github.com/kubernetes/kubernetes/pull/36779

There really is no reason to keep volumes for terminated pods attached on the node.  This PR extends the removal of volumes on the node from memory-backed (the current policy) to all volumes.

@pmorie raised a concern an impact debugging volume related issues if terminated pod volumes are removed.  To address this issue, the PR adds a `--keep-terminated-pod-volumes` flag the kubelet and sets it for `hack/local-up-cluster.sh`.

For consideration in 1.6.

Fixes #35406

@derekwaynecarr @vishh @dashpole

```release-note
kubelet tears down pod volumes on pod termination rather than pod deletion
```

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37228)

### CODE REVIEW

**Comment by smarterclayton** [comment]
In the long run I could imagine tagging specific containers and pods as having a "retainPolicy" that allows debugging of other aspects, not just volumes, post pod termination.  Things like logs, network artifacts, etc.  But that doesn't block or inhibit this work.

**Comment by derekwaynecarr** [comment]
In the long run, i want to be able to schedule on local disk, so keeping disk cleanup with pod lifetime is important.  I guess we could do something like retainPolicy, but it would need to be visible to scheduler.

**Comment by mikedanese** [comment]
@k8s-bot bazel test this, issue kubernetes/test-infra#1152

**Comment by sjenning** [comment]
@saad-ali @kubernetes/sig-storage 

I would like to get this in early in the 1.6 dev cycle so it has time to soak.  Reviews appreciated!

**Comment by vishh** [comment]
cc @dashpole who is working on cleaning up all volumes as part of pod termination.

**Comment by saad-ali** [comment]
> I would like to get this in early in the 1.6 dev cycle so it has time to soak. Reviews appreciated!

Ack, will take a look as soon as 1.5 release is out.

**Comment by derekwaynecarr** [comment]
@vishh -- you agree we could stage the work?  for example, accept this change before taking any larger refactor that @dashpole was looking at performing?

**Comment by derekwaynecarr** [comment]
for example, it seems like we will need to have a flag like this to preserve legacy behavior no matter any future solution

**Comment by vishh** [comment]
Staging is fine by me.

**Comment by sjenning** [comment]
@saad-ali can I get your review on this?  I'd like to give it maximum soak time in the 1.6 dev cycle if you approve of the change.

**Comment by saad-ali** [comment]
> @saad-ali can I get your review on this? I'd like to give it maximum soak time in the 1.6 dev cycle if you approve of the change.

Reviewing

File: cmd/kubelet/app/options/options.go [hunk 1]
```diff
@@ -253,6 +253,7 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&s.ExperimentalKernelMemcgNotification, "experimental-kernel-memcg-notification", s.ExperimentalKernelMemcgNotification, "If enabled, the kubelet will integrate with the kernel memcg notification to determine if memory eviction thresholds are crossed rather than polling.")
 	fs.Int32Var(&s.PodsPerCore, "pods-per-core", s.PodsPerCore, "Number of Pods per core that can run on this Kubelet. The total number of Pods on this Kubelet cannot exceed max-pods, so max-pods will be used if this calculation results in a larger number of Pods allowed on the Kubelet. A value of 0 disables this limit.")
 	fs.BoolVar(&s.ProtectKernelDefaults, "protect-kernel-defaults", s.ProtectKernelDefaults, "Default kubelet behaviour for kernel tuning. If set, kubelet errors if any of kernel tunables is different than kubelet defaults.")
+	fs.BoolVar(&s.KeepTerminatedPodVolumes, "keep-terminated-pod-volumes", s.KeepTerminatedPodVolumes, "Keep terminated pod volumes attached to the node afterh the pod terminates.  Can be useful for debugging volume related issues.")
```
Comment (saad-ali) [inline cmd/kubelet/app/options/options.go]: What is the default for this flag, and where is it set?

File: hack/local-up-cluster.sh [hunk 2]
```diff
@@ -537,6 +537,7 @@ function start_kubelet {
         --experimental-cgroups-per-qos=${EXPERIMENTAL_CGROUPS_PER_QOS} \
         --cgroup-driver=${CGROUP_DRIVER} \
         --cgroup-root=${CGROUP_ROOT} \
+        --keep-terminated-pod-volumes=true \
```
Comment (saad-ali) [inline hack/local-up-cluster.sh]: Why do this for local cluster up?

File: pkg/apis/componentconfig/types.go [hunk 3]
```diff
@@ -483,6 +483,9 @@ type KubeletConfiguration struct {
 	// (binaries, etc.) to mount the volume are available on the underlying node. If the check is enabled
 	// and fails the mount operation fails.
 	ExperimentalCheckNodeCapabilitiesBeforeMount bool
+	// This flag, if set, instructs the kubelet to keep volumes from terminated pods attached to the node.
```
Comment (saad-ali) [inline pkg/apis/componentconfig/types.go]: `attached` -> `mounted`
Ditto in other files

**Review by saad-ali — COMMENTED** [review]
A few minor comments. Otherwise LGTM

File: hack/local-up-cluster.sh (hunk 2 above)
Comment (sjenning) [inline hack/local-up-cluster.sh]: I figured `local-up-cluster.sh` is used for deploying dev environments where it is more likely the user will want to keep the volumes around for debug purposes.  I know that is what I would want personally.

**Comment by jingxu97** [comment]
@sjenning #36779 addressed the pod termination case on kubelet side, should we also do the same on master attach_detach_controller?

File: cmd/kubelet/app/options/options.go (hunk 1 above)
Comment (sjenning) [inline cmd/kubelet/app/options/options.go]: The bool flags are defaulted to 0 automatically by Go afaict.  There are some bool flags that use pointers in v1alpha1 and make the distinction between true/false/unset cases.  But it this case, false is the desired default, and we don't care about tracking the whether the flag was explicitly set false or no present at all (same effect).

**Comment by sjenning** [comment]
@saad-ali I think I addressed your feedback in the latest commits.  Can you look again?

@jingxu97 This is the first I'm learning of a controller the cares about volumes.  I see from https://github.com/kubernetes/kubernetes/issues/20262 this was created to mount cloud volumes (GPD, EBS, Cinder, etc) to the nodes.  Looks like a rough copy of the kubelet volume manager.  Not sure how the states of the world sync between the two.

Does anything need to be done wrt the attachdeatch controller?

**Comment by saad-ali** [comment]
Feel free to readd LGTM label after rebase. And please add a release note.

**Comment by sjenning** [comment]
can I re-add an lgtm on my own PR?

/lgtm

**Pull request merged**